src/sentry/static/sentry/app/utils/withRepositories.tsx

```diff
@@ -0,0 +1,78 @@
+import React from 'react';
+import Reflux from 'reflux';
+import createReactClass from 'create-react-class';
+
+import {Client} from 'app/api';
+import {Repository} from 'app/types';
+import getDisplayName from 'app/utils/getDisplayName';
+import RepositoryStore from 'app/stores/repositoryStore';
+import {getRepositories} from 'app/actionCreators/repositories';
+
+type DependentProps = {
+  api: Client;
+  orgSlug: string;
+  projectSlug: string;
+  releaseVersion: string;
+};
+
+type InjectedProps = {
+  repos?: Array<Repository>;
+  reposLoading?: boolean;
+  reposError?: Error;
+};
+
+const withRepositories = <P extends InjectedProps>(
+  WrappedComponent: React.ComponentType<P>
+) =>
+  createReactClass<
+    Omit<P, keyof InjectedProps> & Partial<InjectedProps> & DependentProps,
+    InjectedProps
+  >({
+    displayName: `withRepositories(${getDisplayName(WrappedComponent)})`,
+    mixins: [Reflux.listenTo(RepositoryStore, 'onStoreUpdate') as any],
+
+    getInitialState() {
+      const {orgSlug} = this.props as P & DependentProps;
+      const repoData = RepositoryStore.get(orgSlug);
+
+      return {...repoData};
+    },
+
+    componentDidMount() {
+      this.fetchRepos();
+    },
+
+    fetchRepos() {
+      const {api, orgSlug} = this.props as P & DependentProps;
+      const repoData = RepositoryStore.get(orgSlug);
+
+      // Case 1: Data has not loaded
+      // Case 2: Data is older than 30s
+      if (
+        (!repoData.repos && !repoData.reposLoading) ||
+        Date.now() - RepositoryStore.lastUpdated > 30000
```

Comment:
We are trying to reduce the quantity/frequency of API calls to the repo endpoint, while still ensuring that data inside the store stays "fresh". 

30s is an arbitrarily chosen number that seems short enough that isn't bad UX, but long enough that the cache still serves its function.

Reply:
Are we not able to flush/clear the store when repositories are changed in settings?

Reply:
Yeah, how often are repositories changed?

Reply:
Expiring data using time is objectively bad and I've reflected upon my sins and made a better implementation to redeem my soul.

Reply:
I made adding/modifying repositories inside settings to fire an event to reset the store.